### PR TITLE
[sdk.logs] Fix the version reported by LogRecord.Logger when using ILogger integration

### PR DESCRIPTION
**Pull request by CodeBlanch**
Relates to #4433

## Changes

* Fixes the logger instrumentation scope when using ILogger to report the {majorVersion}.{minorVersion}.{patchVersion}.{pre-release label}.{pre-release version}.{gitHeight} from the SDK `AssemblyInformationalVersionAttribute` (eg `1.5.0-alpha.1`) instead of `AssemblyVersion` (eg `1.0.0.0`).

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated

### CODE REVIEW

File: src/OpenTelemetry/Sdk.cs [hunk 1]
```diff
@@ -41,13 +42,20 @@ static Sdk()
             Activity.DefaultIdFormat = ActivityIdFormat.W3C;
             Activity.ForceDefaultIdFormat = true;
             SelfDiagnostics.EnsureInitialized();
+
+            var sdkVersion = typeof(Sdk).Assembly.GetCustomAttribute<AssemblyFileVersionAttribute>()?.Version;
```
Comment (vishweshbankwar) [inline src/OpenTelemetry/Sdk.cs:46]: Does this include pre-release info? We were including that https://github.com/open-telemetry/opentelemetry-dotnet/pull/4635/files#diff-82711f0e4129a4bdcd2306a09e8ba95ca230d6b911561e8f38461aa242c7b954:~:text=%7D-,private%20static%20string%20GetAssemblyInformationalVersion(),%7D,-%7D

Comment (CodeBlanch) [reply src/OpenTelemetry/Sdk.cs]: Restored! Sorry about that

**Review by vishweshbankwar — APPROVED** [review]

File: src/OpenTelemetry/Sdk.cs (hunk 1 above)
Comment (reyang) [reply src/OpenTelemetry/Sdk.cs]: How do we test this? The current test seems to be insufficient.

File: src/OpenTelemetry/Sdk.cs [hunk 2]
```diff
@@ -41,13 +42,33 @@ static Sdk()
...
+                ? assemblyInformationalVersion.Substring(0, indexOfPlusSign)
+                : assemblyInformationalVersion;
         }
 
         /// <summary>
         /// Gets a value indicating whether instrumentation is suppressed (disabled).
         /// </summary>
         public static bool SuppressInstrumentation => SuppressInstrumentationScope.IsSuppressed;
 
+        internal static string InformationalVersion { get; }
```
Comment (utpilla) [inline src/OpenTelemetry/Sdk.cs:70]:
```suggestion
        internal static readonly string InformationalVersion;
```

File: test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj [hunk 3]
```diff
@@ -16,7 +16,8 @@
   </ItemGroup>
 
   <ItemGroup>
-	<PackageReference Include="Microsoft.Extensions.Logging" />
+    <PackageReference Include="Microsoft.Extensions.Logging" />
+    <PackageReference Include="Microsoft.CSharp" Condition="'$(TargetFramework)' == 'net462'" />
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj:20]: Why is this needed?

File: src/OpenTelemetry/Sdk.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Sdk.cs:70]: I'm not opposed to this, but why? 🤣 A property with just a getter (`{ get; }`) is `readonly` for the backing generated by the compiler IIRC.

File: test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj (hunk 3 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj:20]: These tests were being skipped on `NETFRAMEWORK`: https://github.com/open-telemetry/opentelemetry-dotnet/blob/104cf32d97b2aba7319635fc23d84d5f7663fe9b/test/OpenTelemetry.Tests/Logs/LogRecordTest.cs#L17

I found that odd because we support logging on `net462`+.

Turns out it is because of this one test using `dynamic`: https://github.com/open-telemetry/opentelemetry-dotnet/blob/104cf32d97b2aba7319635fc23d84d5f7663fe9b/test/OpenTelemetry.Tests/Logs/LogRecordTest.cs#L240

Adding `Microsoft.CSharp` for `net462` brings in `dynamic` support.

File: src/OpenTelemetry/Sdk.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Sdk.cs]: To be fully robust probably needs an integration test that looks at the actual dll's AssemblyInformationalVersionAttribute spit out by the build system? Don't want to tackle that on this PR but I added a unit test that covers the parse logic.

File: src/OpenTelemetry/Sdk.cs (hunk 2 above)
Comment (utpilla) [reply src/OpenTelemetry/Sdk.cs:55]: > but why? 

Just because it's more obvious when you use `readonly` 😀

**Review by utpilla — APPROVED** [review]
